action/protocol/poll/staking_command.go

```diff
@@ -81,29 +85,22 @@ func (sc *stakingCommand) Start(ctx context.Context) error {
 }
 
 func (sc *stakingCommand) CreatePreStates(ctx context.Context, sm protocol.StateManager) error {
-	// TODO: handle V2
-	return sc.stakingV1.CreateGenesisStates(ctx, sm)
+	return sc.slasher.CreatePreStates(ctx, sm, sc.candIndexer)
 }
 
 func (sc *stakingCommand) CreatePostSystemActions(ctx context.Context) ([]action.Envelope, error) {
 	return createPostSystemActions(ctx, sc)
 }
 
 func (sc *stakingCommand) Handle(ctx context.Context, act action.Action, sm protocol.StateManager) (*action.Receipt, error) {
-	if sc.stakingV1 == nil {
-		return handle(ctx, act, sm, sc.candIndexer, sc.addr.String())
-	}
-
-	if sc.stakingV2 == nil {
-		return sc.stakingV1.Handle(ctx, act, sm)
-	}
-
 	// transition to V2 starting Fairbank
 	height, err := sm.Height()
 	if err != nil {
 		return nil, err
 	}
-	if sc.hu.IsPost(config.Fairbank, height) {
+	bcCtx := protocol.MustGetBlockchainCtx(ctx)
+	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	if sc.stakingV1 == nil || hu.IsPost(config.Fairbank, height) {
```

Comment:
might be an issue to simplify the logic like this

for stakingV1 == nil, and height < Fairbank, it will skip the if(), and call stakingV1.Handle(ctx, act, sm), that will panic

Reply:
i think the way in ReadState() is correct/better